CONFIGURATION.md
# Configuration Options

The following configuration options are available and can be passed as environment variables.

If you don't know how to set them, look up the section for your type of installment
in the [INSTALL.md](./INSTALL.md).

## General

<!-- markdownlint-disable -->

| Variable              | Description                                                                                                                                                                                                         | Type   | Default Value                |
| --------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------ | ---------------------------- |
| `DASHDOT_PORT`        | The port where the express backend is running (the backend serves the frontend, so it is the same port for both)                                                                                                    | number | `3001`                       |
| `DASHDOT_WIDGET_LIST` | Change the order of the elements in the list, to change the position on the page, or remove an item from the list, to remove it from the page (The available options are: `os`, `cpu`, `storage`, `ram`, `network`) | string | `os,cpu,storage,ram,network` |

<!-- markdownlint-enable -->

## OS Widget

<!-- markdownlint-disable -->

| Variable                      | Description                                                                                                                                                                                           | Type    | Default Value      |
| ----------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------- | ------------------ |
| `DASHDOT_DISABLE_HOST`        | If you want to hide the host part in the server widget (e.g. `dash.mauz.io` -> `dash.`)                                                                                                               | boolean | `false`            |
| `DASHDOT_OS_LABEL_LIST`       | Change the order of the labels in the list, to change the position in the widget, or remove an item from the list, to remove it from the widget (The available options are: `os`, `arch`, `up_since`) | string  | `os,arch,up_since` |
| `DASHDOT_OS_WIDGET_GROW`      | To adjust the relative size of the OS widget                                                                                                                                                          | number  | `1.5`              |
| `DASHDOT_OS_WIDGET_MIN_WIDTH` | To adjust the minimum width of the OS widget (in px)                                                                                                                                                  | number  | `300`              |

<!-- markdownlint-enable -->

## CPU Widget

<!-- markdownlint-disable -->

| Variable                       | Description                                                                                                                                                                                                                    | Type    | Default Value                         |
| ------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------- | ------------------------------------- |
| `DASHDOT_ENABLE_CPU_TEMPS`     | If you want to show the CPU temperature in the graph. This will probably not work on a VPS, so you need to try it on your own if this works. For home servers it might work just fine                                          | boolean | `false`                               |
| `DASHDOT_CPU_LABEL_LIST`       | Change the order of the labels in the list, to change the position in the widget, or remove an item from the list, to remove it from the widget (The available options are: `brand`, `model`, `cores`, `threads`, `frequency`) | string  | `brand,model,cores,threads,frequency` |
| `DASHDOT_CPU_WIDGET_GROW`      | To adjust the relative size of the Processor widget                                                                                                                                                                            | number  | `4`                                   |
| `DASHDOT_CPU_WIDGET_MIN_WIDTH` | To adjust the minimum width of the Processor widget (in px)                                                                                                                                                                    | number  | `500`                                 |
| `DASHDOT_CPU_DATAPOINTS`       | The amount of datapoints in the Processor graph                                                                                                                                                                                | number  | `20`                                  |
| `DASHDOT_CPU_POLL_INTERVAL`    | Read the Processor load every x milliseconds                                                                                                                                                                                   | number  | `1000`                                |

<!-- markdownlint-enable -->

## Storage Widget

<!-- markdownlint-disable -->

| Variable                           | Description                                                                                                                                                                                          | Type   | Default Value     |
| ---------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------ | ----------------- |
| `DASHDOT_STORAGE_LABEL_LIST`       | Change the order of the labels in the list, to change the position in the widget, or remove an item from the list, to remove it from the widget (The available options are: `brand`, `size`, `type`) | string | `brand,size,type` |
| `DASHDOT_STORAGE_WIDGET_GROW`      | To adjust the relative size of the Storage widget                                                                                                                                                    | number | `3.5`             |
| `DASHDOT_STORAGE_WIDGET_MIN_WIDTH` | To adjust the minimum width of the Storage widget (in px)                                                                                                                                            | number | `500`             |
| `DASHDOT_STORAGE_POLL_INTERVAL`    | Read the Storage load every x milliseconds                                                                                                                                                           | number | `60000`           |

<!-- markdownlint-enable -->

## RAM Widget

<!-- markdownlint-disable -->

| Variable                       | Description                                                                                                                                                                                                       | Type   | Default Value               |
| ------------------------------ | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------ | --------------------------- |
| `DASHDOT_RAM_LABEL_LIST`       | Change the order of the labels in the list, to change the position in the widget, or remove an item from the list, to remove it from the widget (The available options are: `brand`, `size`, `type`, `frequency`) | string | `brand,size,type,frequency` |
| `DASHDOT_RAM_WIDGET_GROW`      | To adjust the relative size of the Memory widget                                                                                                                                                                  | number | `4`                         |
| `DASHDOT_RAM_WIDGET_MIN_WIDTH` | To adjust the minimum width of the Memory widget (in px)                                                                                                                                                          | number | `500`                       |
| `DASHDOT_RAM_DATAPOINTS`       | The amount of datapoints in the Memory graph                                                                                                                                                                      | number | `20`                        |
| `DASHDOT_RAM_POLL_INTERVAL`    | Read the Memory load every x milliseconds                                                                                                                                                                         | number | `1000`                      |

<!-- markdownlint-enable -->

## Network Widget

<!-- markdownlint-disable -->

| Variable                           | Description                                                                                                                                                                                                                                   | Type   | Default Value                              |
| ---------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------ | ------------------------------------------ |
| `DASHDOT_SPEED_TEST_INTERVAL`      | At which interval the network speed-test should be rerun (in minutes)                                                                                                                                                                         | number | `60`                                       |
| `DASHDOT_NETWORK_LABEL_LIST`       | Change the order of the labels in the list, to change the position in the widget, or remove an item from the list, to remove it from the widget (The available options are: `type`, `speed_up`, `speed_down`, `interface_speed`, `public_ip`) | string | `type,speed_up,speed_down,interface_speed` |
| `DASHDOT_NETWORK_WIDGET_GROW`      | To adjust the relative size of the Network widget                                                                                                                                                                                             | number | `6`                                        |
| `DASHDOT_NETWORK_WIDGET_MIN_WIDTH` | To adjust the minimum width of the Network widget (in px)                                                                                                                                                                                     | number | `500`                                      |
| `DASHDOT_NETWORK_DATAPOINTS`       | The amount of datapoints in each of the Network graphs                                                                                                                                                                                        | number | `20`                                       |
| `DASHDOT_NETWORK_POLL_INTERVAL`    | Read the Network load every x milliseconds                                                                                                                                                                                                    | number | `1000`                                     |

<!-- markdownlint-enable -->

## Overrides

Override specific fields, by providing your desired value with the following options.

<!-- markdownlint-disable -->

| Variable                                   | Description                                                                                                                                                           | Type   | Default Value |
| ------------------------------------------ | --------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------ | ------------- |
| `DASHDOT_OVERRIDE_OS`                      |                                                                                                                                                                       | string |
| `DASHDOT_OVERRIDE_ARCH`                    |                                                                                                                                                                       | string |
| `DASHDOT_OVERRIDE_CPU_BRAND`               |                                                                                                                                                                       | string |
| `DASHDOT_OVERRIDE_CPU_MODEL`               |                                                                                                                                                                       | string |
| `DASHDOT_OVERRIDE_CPU_CORES`               |                                                                                                                                                                       | number |
| `DASHDOT_OVERRIDE_CPU_THREADS`             |                                                                                                                                                                       | number |
| `DASHDOT_OVERRIDE_CPU_FREQUENCY`           | Number needs to be passed in GHz (e.g. `2.8`)                                                                                                                         | number |
| `DASHDOT_OVERRIDE_RAM_BRAND`               |                                                                                                                                                                       | string |
| `DASHDOT_OVERRIDE_RAM_SIZE`                | Number needs to be passed in bytes (e.g. `34359738368` for 32 GB, because it is `32 * 1024 * 1024 * 1024`)                                                            | number |
| `DASHDOT_OVERRIDE_RAM_TYPE`                |                                                                                                                                                                       | string |
| `DASHDOT_OVERRIDE_RAM_FREQUENCY`           |                                                                                                                                                                       | number |
| `DASHDOT_OVERRIDE_NETWORK_TYPE`            |                                                                                                                                                                       | string |
| `DASHDOT_OVERRIDE_NETWORK_SPEED_UP`        | Number needs to be passed in bit (e.g. `100000000` for 100 Mb/s, because it is `100 * 1000 * 1000`)                                                                   | number |
| `DASHDOT_OVERRIDE_NETWORK_SPEED_DOWN`      | Number needs to be passed in bit (e.g. `100000000` for 100 Mb/s, because it is `100 * 1000 * 1000`)                                                                   | number |
| `DASHDOT_OVERRIDE_NETWORK_INTERFACE_SPEED` | Number needs to be passed in Megabit (e.g. `10000` for 10 GB/s, because it is `10 * 1000`)                                                                            | number |
| `DASHDOT_OVERRIDE_NETWORK_PUBLIC_IP`       |                                                                                                                                                                       | string |
| `DASHDOT_OVERRIDE_STORAGE_BRAND_[1-5]`     | Use a suffix of 1, 2, 3, 4 or 5 for the respective drives                                                                                                             | string |
| `DASHDOT_OVERRIDE_STORAGE_SIZE_[1-5]`      | Use a suffix of 1, 2, 3, 4 or 5 for the respective drives. Number needs to be passed in bytes (e.g. `34359738368` for 32 GB, because it is `32 * 1024 * 1024 * 1024`) | number |
| `DASHDOT_OVERRIDE_STORAGE_TYPE_[1-5]`      | Use a suffix of 1, 2, 3, 4 or 5 for the respective drives                                                                                                             | string |

<!-- markdownlint-enable -->


CHANGELOG.md
## [3.3.3](https://github.com/MauriceNino/dashdot/compare/v3.3.2...v3.3.3) (2022-06-13)


### Bug Fixes

* **deps:** upgrade systeminformation to v5.11.20 ([404ed9e](https://github.com/MauriceNino/dashdot/commit/404ed9ed9dea7cda9f9e0f4e6ef93554019f9d91)), closes [#111](https://github.com/MauriceNino/dashdot/issues/111) [#114](https://github.com/MauriceNino/dashdot/issues/114)

## [3.3.2](https://github.com/MauriceNino/dashdot/compare/v3.3.1...v3.3.2) (2022-06-13)


### Bug Fixes

* **view:** missing box-shadow on charts (firefox) ([02bf98b](https://github.com/MauriceNino/dashdot/commit/02bf98bf6ab52e409b86e4366654dd221b95ba01))

## [3.3.1](https://github.com/MauriceNino/dashdot/compare/v3.3.0...v3.3.1) (2022-06-13)


### Bug Fixes

* **deps:** revert - upgrade systeminformation to v5.11.17 ([41f2150](https://github.com/MauriceNino/dashdot/commit/41f2150e401344f66b74ba2f0cc4078c4fbf8632)), closes [#114](https://github.com/MauriceNino/dashdot/issues/114)

# [3.3.0](https://github.com/MauriceNino/dashdot/compare/v3.2.1...v3.3.0) (2022-06-13)


### Bug Fixes

* **api,docker:** fixed speedtest bugs ([dbfbc68](https://github.com/MauriceNino/dashdot/commit/dbfbc68fef77bf1296378e82db96722b46e77040))
* **api,docker:** fixed speedtest install ([2ae1be2](https://github.com/MauriceNino/dashdot/commit/2ae1be2ed0f968e3a5b01b9d88e0067bf5af24d8))
* **api,docker:** replaced speedtest-cli with speedtest ([3f1b5ac](https://github.com/MauriceNino/dashdot/commit/3f1b5ac7bc273455e38815beaae7442a6aca238f))
* **api:** fixed release for Arch Linux ([168c759](https://github.com/MauriceNino/dashdot/commit/168c759a8bf52862d583d84643c339e143f137b8))


### Features

* **api:** add log output for speedtest runner ([26267b2](https://github.com/MauriceNino/dashdot/commit/26267b2668d90648c815938e313a8577b5d3bba5))
* **cli:** add buildhash and version to info output ([c683ee4](https://github.com/MauriceNino/dashdot/commit/c683ee4c5969334b51b9abe9a100635cd7413023))
* **cli:** add cli to better handle issues ([0c7944f](https://github.com/MauriceNino/dashdot/commit/0c7944fc55ed6cfb87397da4243e20555f06acd8))

## [3.2.1](https://github.com/MauriceNino/dashdot/compare/v3.2.0...v3.2.1) (2022-06-12)


### Bug Fixes

* **api:** invalid brand & model ([51f593e](https://github.com/MauriceNino/dashdot/commit/51f593e6b3b5e2f4c20f7ee43159ee83ccb6808d))

# [3.2.0](https://github.com/MauriceNino/dashdot/compare/v3.1.3...v3.2.0) (2022-06-12)


### Bug Fixes

* **api:** calculate network speed in relation to time diff ([55ec6b2](https://github.com/MauriceNino/dashdot/commit/55ec6b22d0668a5f31ea3a959586d35423a2ed5c))
* **api:** match any mounts that are present at /mnt/host_* ([810e5a6](https://github.com/MauriceNino/dashdot/commit/810e5a6dac7e71ee28fd388bb0570d0378bba7ea)), closes [#108](https://github.com/MauriceNino/dashdot/issues/108)


### Features

* **view:** show error widget when static data cant be loaded ([89d9fcc](https://github.com/MauriceNino/dashdot/commit/89d9fcc3ab976fc39e40f53a271984453298c0bd))

## [3.1.3](https://github.com/MauriceNino/dashdot/compare/v3.1.2...v3.1.3) (2022-06-09)


### Bug Fixes

* reduced docker image size ([d29ed54](https://github.com/MauriceNino/dashdot/commit/d29ed546aa9b4273191c671282a5c1d9873e9d34))

## [3.1.2](https://github.com/MauriceNino/dashdot/compare/v3.1.1...v3.1.2) (2022-06-09)


### Bug Fixes

* **api:** read temperature from main if no per-core data is found ([523e7a1](https://github.com/MauriceNino/dashdot/commit/523e7a1e768e55886832c6fe876dcd138d12e577)), closes [#107](https://github.com/MauriceNino/dashdot/issues/107)
* **view:** error on domains with multiple subdomains ([756774b](https://github.com/MauriceNino/dashdot/commit/756774b9ecf54b76bd074270e737c122447b7fb9))
* **view:** mobile browser address bar not hiding on scroll ([7a32a7c](https://github.com/MauriceNino/dashdot/commit/7a32a7cdaf478009d3c08b51231720e8660635f1))
* **view:** remove old CRA app name in manifest ([dd8e0a6](https://github.com/MauriceNino/dashdot/commit/dd8e0a617a1a4837d41103560b53b0b604206188))
* **view:** server image not shown on firefox ([adb6607](https://github.com/MauriceNino/dashdot/commit/adb660795ab51ddf739b5c0b931b138ab1555994))

## [3.1.1](https://github.com/MauriceNino/dashdot/compare/v3.1.0...v3.1.1) (2022-06-09)


### Bug Fixes

* **api:** invalid iface speed conversion for non-docker interface ([2a1425f](https://github.com/MauriceNino/dashdot/commit/2a1425f7fd402a006855b8cbb53cd9f8d1aedfe0)), closes [#105](https://github.com/MauriceNino/dashdot/issues/105)

# [3.1.0](https://github.com/MauriceNino/dashdot/compare/v3.0.0...v3.1.0) (2022-06-08)


### Bug Fixes

* **api:** exclude /etc from mounts ([bb0c2aa](https://github.com/MauriceNino/dashdot/commit/bb0c2aa44c3d933f158ef711ec0ed1f9d3448cd9))
* **api:** get sizes of all drives instead of just the main one ([356cacd](https://github.com/MauriceNino/dashdot/commit/356cacd47ac4f947c93e9a216dc5fbc05de8bbcb)), closes [#59](https://github.com/MauriceNino/dashdot/issues/59)
* **api:** read type and interface speed from host instead of container ([2f0eea4](https://github.com/MauriceNino/dashdot/commit/2f0eea45d4583464b5429e4f6c3d12b56cb5d645)), closes [#105](https://github.com/MauriceNino/dashdot/issues/105)
* **view:** extract label function to widget ([8718f5e](https://github.com/MauriceNino/dashdot/commit/8718f5e86d13ec9157f1b18baed77f1bdea5ce37))


### Features

* **view:** add percentage to storage widget ([e2cc3fa](https://github.com/MauriceNino/dashdot/commit/e2cc3fad305284e4ddd382958b6bb74fb0aedd0b)), closes [#104](https://github.com/MauriceNino/dashdot/issues/104)

# [3.0.0](https://github.com/MauriceNino/dashdot/compare/v2.3.0...v3.0.0) (2022-06-04)


### Bug Fixes

* **ci:** add correct drone hash ([c3ecf2c](https://github.com/MauriceNino/dashdot/commit/c3ecf2c18e12f880bfc6f8e2d6a2c4c77085a191))

apps/api/src/config.ts
import { Config } from '@dash/common';

const numNull = (val: string | undefined): number | undefined => {
  if (val === undefined) {
    return undefined;
  }
  return +val;
};

const penv = (key: string): string | undefined => process.env[`DASHDOT_${key}`];
const lst = (item: string): any[] => item.split(',');

export const CONFIG: Config = {
  port: numNull(penv('PORT')) ?? 3001,
  widget_list: lst(penv('WIDGET_LIST') ?? 'os,cpu,storage,ram,network'),
  accept_ookla_eula: penv('ACCEPT_OOKLA_EULA') === 'true',

  disable_host: penv('DISABLE_HOST') === 'true',
  os_label_list: lst(penv('OS_LABEL_LIST') ?? 'os,arch,up_since'),
  os_widget_grow: numNull(penv('OS_WIDGET_GROW')) ?? 1.5,
  os_widget_min_width: numNull(penv('OS_WIDGET_MIN_WIDTH')) ?? 300,

  enable_cpu_temps: penv('ENABLE_CPU_TEMPS') === 'true',
  cpu_label_list: lst(
    penv('CPU_LABEL_LIST') ?? 'brand,model,cores,threads,frequency'
  ),
  cpu_widget_grow: numNull(penv('CPU_WIDGET_GROW')) ?? 4,
  cpu_widget_min_width: numNull(penv('CPU_WIDGET_MIN_WIDTH')) ?? 500,
  cpu_shown_datapoints: numNull(penv('CPU_SHOWN_DATAPOINTS')) ?? 20,
  cpu_poll_interval: numNull(penv('CPU_POLL_INTERVAL')) ?? 1000,

  storage_label_list: lst(penv('STORAGE_LABEL_LIST') ?? 'brand,size,type'),
  storage_widget_grow: numNull(penv('STORAGE_WIDGET_GROW')) ?? 3.5,
  storage_widget_min_width: numNull(penv('STORAGE_WIDGET_MIN_WIDTH')) ?? 500,
  storage_poll_interval: numNull(penv('STORAGE_POLL_INTERVAL')) ?? 60000,

  ram_label_list: lst(penv('RAM_LABEL_LIST') ?? 'brand,size,type,frequency'),
  ram_widget_grow: numNull(penv('RAM_WIDGET_GROW')) ?? 4,
  ram_widget_min_width: numNull(penv('RAM_WIDGET_MIN_WIDTH')) ?? 500,
  ram_shown_datapoints: numNull(penv('RAM_SHOWN_DATAPOINTS')) ?? 20,
  ram_poll_interval: numNull(penv('RAM_POLL_INTERVAL')) ?? 1000,

  speed_test_interval: numNull(penv('SPEED_TEST_INTERVAL')) ?? 60,
  network_label_list: lst(
    penv('NETWORK_LABEL_LIST') ?? 'type,speed_up,speed_down,interface_speed'
  ),
  network_widget_grow: numNull(penv('NETWORK_WIDGET_GROW')) ?? 6,
  network_widget_min_width: numNull(penv('NETWORK_WIDGET_MIN_WIDTH')) ?? 500,
  network_shown_datapoints: numNull(penv('NETWORK_SHOWN_DATAPOINTS')) ?? 20,
  network_poll_interval: numNull(penv('NETWORK_POLL_INTERVAL')) ?? 1000,

  override: {
    os: penv('OVERRIDE_OS'),
    arch: penv('OVERRIDE_ARCH'),
    cpu_brand: penv('OVERRIDE_CPU_BRAND'),
    cpu_model: penv('OVERRIDE_CPU_MODEL'),
    cpu_cores: numNull(penv('OVERRIDE_CPU_CORES')),
    cpu_threads: numNull(penv('OVERRIDE_CPU_THREADS')),
    cpu_frequency: numNull(penv('OVERRIDE_CPU_FREQUENCY')),
    ram_brand: penv('OVERRIDE_RAM_BRAND'),
    ram_size: numNull(penv('OVERRIDE_RAM_SIZE')),
    ram_type: penv('OVERRIDE_RAM_TYPE'),
    ram_frequency: numNull(penv('OVERRIDE_RAM_FREQUENCY')),
    network_type: penv('OVERRIDE_NETWORK_TYPE'),
    network_speed_up: numNull(penv('OVERRIDE_NETWORK_SPEED_UP')),
    network_speed_down: numNull(penv('OVERRIDE_NETWORK_SPEED_DOWN')),
    network_interface_speed: numNull(penv('OVERRIDE_NETWORK_INTERFACE_SPEED')),
    network_public_ip: penv('OVERRIDE_NETWORK_PUBLIC_IP'),
    storage_brand_1: penv('OVERRIDE_STORAGE_BRAND_1'),
    storage_size_1: numNull(penv('OVERRIDE_STORAGE_SIZE_1')),
    storage_type_1: penv('OVERRIDE_STORAGE_TYPE_1'),
    storage_brand_2: penv('OVERRIDE_STORAGE_BRAND_2'),
    storage_size_2: numNull(penv('OVERRIDE_STORAGE_SIZE_2')),
    storage_type_2: penv('OVERRIDE_STORAGE_TYPE_2'),
    storage_brand_3: penv('OVERRIDE_STORAGE_BRAND_3'),
    storage_size_3: numNull(penv('OVERRIDE_STORAGE_SIZE_3')),
    storage_type_3: penv('OVERRIDE_STORAGE_TYPE_3'),
    storage_brand_4: penv('OVERRIDE_STORAGE_BRAND_4'),
    storage_size_4: numNull(penv('OVERRIDE_STORAGE_SIZE_4')),
    storage_type_4: penv('OVERRIDE_STORAGE_TYPE_4'),
    storage_brand_5: penv('OVERRIDE_STORAGE_BRAND_5'),
    storage_size_5: numNull(penv('OVERRIDE_STORAGE_SIZE_5')),
    storage_type_5: penv('OVERRIDE_STORAGE_TYPE_5'),
  },
};
